pkg/csi/service/common/authmanager.go

```diff
@@ -0,0 +1,179 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	cnsvsphere "sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+)
+
+// AuthorizationService exposes interfaces to support authorization check on datastores and get datastores which need to be
+// ignored by create volume
+type AuthorizationService interface {
+
+	// GetDatastoreIgnoreMapForBlockVolumes returns a map which maps datastore url to datastore info which user does not have Datastore.FileManagement privilege
+	// returns an error if not able to retrieve the map
+	GetDatastoreIgnoreMapForBlockVolumes(ctx context.Context) (map[string]*cnsvsphere.DatastoreInfo, error)
+}
+
+// AuthManager maintains an internal map to track the datastores that need to be ignored by create volume
+type AuthManager struct {
+	// map the datastore url to datastore info which need to be ignored when invoking CNS CreateVolume API
+	datastoreIgnoreMapForBlockVolumes map[string]*cnsvsphere.DatastoreInfo
+	// this mutex is to make sure the update for datatoreIgnoreMap is mutually exclusive
+	rwMutex sync.RWMutex
+	// vCenter Instance
+	vcenter *cnsvsphere.VirtualCenter
+}
+
+// onceForAuthorizationService is used for initializing the AuthorizationService singleton.
+var onceForAuthorizationService sync.Once
+
+// authManagerIntance is instance of authManager and implements interface for AuthorizationService
+var authManagerInstance *AuthManager
+
+// GetAuthorizationService returns the singleton AuthorizationService
+func GetAuthorizationService(ctx context.Context, vc *cnsvsphere.VirtualCenter) (AuthorizationService, error) {
+	log := logger.GetLogger(ctx)
+	onceForAuthorizationService.Do(func() {
+		log.Info("Initializing authorization service...")
+
+		authManagerInstance = &AuthManager{
+
+			datastoreIgnoreMapForBlockVolumes: make(map[string]*cnsvsphere.DatastoreInfo),
+			rwMutex:                           sync.RWMutex{},
+			vcenter:                           vc,
+		}
+		authManagerInstance.refreshDatastoreIgnoreMap()
+	})
+
+	log.Info("authorization service initialized")
+	return authManagerInstance, nil
+}
+
+// GetDatastoreIgnoreMapForBlockVolumes returns a DatastoreIgnoreMapForBlockVolumes. This map maps datastore url to datastore info
+// which need to be ignored when creating block volume.
+func (authManager *AuthManager) GetDatastoreIgnoreMapForBlockVolumes(ctx context.Context) (map[string]*cnsvsphere.DatastoreInfo, error) {
```

Comment:
Do we need to return error from this function? I see that we are always returning nil.

Reply:
Yes. Currently it always return nil. But I think it is better to have an error in the return value. Maybe later this function need return some error.

Reply:
Its not a good practice to return error when there is no current logic for that. Can you remove it?